examples/basic_bitcoin/src/backend/bitcoin_wallet.ts

```diff
@@ -0,0 +1,295 @@
+//! * Support for address types that aren't P2PKH.
+//! * Caching spent UTXOs so that they are not reused in future transactions.
+//! * Option to set the fee.
+import {
+    BitcoinNetwork,
+    MillisatoshiPerByte,
+    Satoshi,
+    Utxo
+} from 'azle/canisters/management';
+import * as bitcoin from 'bitcoinjs-lib';
+import { networks, Psbt, Transaction } from 'bitcoinjs-lib';
+import { ValidateSigFunction } from 'bitcoinjs-lib/src/psbt';
+import { Buffer } from 'buffer';
+
+import * as bitcoinApi from './bitcoin_api';
+import * as ecdsaApi from './ecdsa_api';
+import { BitcoinTxid, DerivationPath } from './types';
+
+type TransactionHashes = {
+    [txid: string]: string;
+};
+
+/// Returns the P2PKH address of this canister at the given derivation path.
+export async function getP2pkhAddress(
+    network: BitcoinNetwork,
+    keyName: string,
+    derivationPath: DerivationPath
+): Promise<string> {
+    // Fetch the public key of the given derivation path.
+    const publicKey = await ecdsaApi.ecdsaPublicKey(keyName, derivationPath);
+
+    // Compute the address.
+    return publicKeyToP2pkhAddress(network, publicKey);
+}
+
+/// Sends a transaction to the network that transfers the given amount to the
+/// given destination, where the source of the funds is the canister itself
+/// at the given derivation path.
+export async function send(
+    network: BitcoinNetwork,
+    derivationPath: DerivationPath,
+    keyName: string,
+    dstAddress: string,
+    amount: Satoshi,
+    transactionHashes: TransactionHashes
+): Promise<BitcoinTxid> {
+    // Get fee percentiles from previous transactions to estimate our own fee.
+    const feePercentiles = await bitcoinApi.getCurrentFeePercentiles(network);
+
+    const feePerByte =
+        feePercentiles.length === 0
+            ? // There are no fee percentiles. This case can only happen on a regtest
+              // network where there are no non-coinbase transactions. In this case,
+              // we use a default of 2000 millisatoshis/byte (i.e. 2 satoshi/byte)
+              2_000n
+            : // Choose the 50th percentile for sending fees.
+              feePercentiles[49];
+
+    // Fetch our public key, P2PKH address, and UTXOs.
+    const ownAddress = await getP2pkhAddress(network, keyName, derivationPath);
+    const ownPublicKey = await ecdsaApi.ecdsaPublicKey(keyName, derivationPath);
+
+    console.log('Fetching UTXOs...');
+    // Note that pagination may have to be used to get all UTXOs for the given address.
+    // For the sake of simplicity, it is assumed here that the `utxo` field in the response
+    // contains all UTXOs.
+    const ownUtxos = (await bitcoinApi.getUtxos(network, ownAddress.toString()))
+        .utxos;
+
+    // Build the transaction that sends `amount` to the destination address.
+    const transaction = await buildTransaction(
+        ownPublicKey,
+        ownAddress,
+        ownUtxos,
+        dstAddress,
+        amount,
+        feePerByte,
```

Comment:
We do we have to pass in all of these extra arguments?

Reply:
Because psbt needs it, I moved to regular transactions and that cleared up most of them